Add documentation for default resiliency policies

halspang opened the pull request:
## Description

This commit adds an overview of the generic policies, their keywords, and a brief example of how they are chosen.

https://github.com/dapr/docs/issues/2664

Signed-off-by: Hal Spang <halspang@microsoft.com>

## Issue reference

#2664

halspang committed:
Add documentation for generic resiliency policies

This commit adds an overview of the generic policies, their keywords,
and a brief example of how they are chosen.

https://github.com/dapr/docs/issues/2664

Signed-off-by: Hal Spang <halspang@microsoft.com>

greenie-msft commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +132,64 @@ spec:
+
+#### Utilize Generic Policies
+
+Dapr Resiliency also lets you define policies that can have a broader scope than the standard ones defined above. This is done through the usage of keywords that let Dapr know when to apply the given policy. There are 3 base policies:
```
```suggestion
In resiliency you can set default policies, which can have a broader scope. This is done through reserved keywords that let Dapr know when to apply the given policy. There are 3 default policies types: 
```

greenie-msft commented on daprdocs/content/en/operations/resiliency/policies.md:
````diff
@@ -132,4 +132,64 @@ spec:
\ No newline at end of file
+```
+
+#### Utilize Generic Policies
````
```suggestion
#### Setting Default Policies
```

greenie-msft commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +132,64 @@ spec:
+| ConfigurationComponentOutbound | All configuration component operations.              | DefaultConfigurationComponentOutboundPolicy | 
+| LockComponentOutbound          | All lock component operations.                       | DefaultLockComponentOutboundPolicy          | 
+
+For all generic policies, they will only be used if the operation being executed matches the policy and if there is no more specific policy targeting it.
```
```suggestion
Default policies are applied if the operation being executed matches the policy type and if there is no more specific policy targeting it.
```

greenie-msft commented on daprdocs/content/en/operations/resiliency/policies.md:
````diff
@@ -132,4 +132,64 @@ spec:
+        retry: retryForever
+```
+
+In this scenario, when AppA is called, the `fastRetries` policy will be used. For AppB, `retryForever` will be used. Finally, when calling AppC, `DefaultRetryPolicy` will be called even though it was never applied via a target.
````
```suggestion
In this scenario, when AppA is called, the `fastRetries` policy is used. For AppB, `retryForever` is used. Finally, when calling AppC, `DefaultRetryPolicy` is called even though it was never applied to a target.
```

greenie-msft commented:
@halspang can you also describe the behavior if I have two default policies applied - for example, if I have both DefaultAppRetry is defined as well as DefaultRetry, which one wins? I assume DefaultAppRetry does, but it would be good to call that out.

It will be good to show this in at least two or three more example - the hierarchy of policies and the scenarios for when they are applied.

greenie-msft commented:
ALso @halspang the table should include defaults. I know we have default for app to app. What about the other defaults?

halspang committed:
Update daprdocs/content/en/operations/resiliency/policies.md

Co-authored-by: greenie-msft <56556602+greenie-msft@users.noreply.github.com>
Signed-off-by: halspang <70976921+halspang@users.noreply.github.com>

halspang committed:
Update daprdocs/content/en/operations/resiliency/policies.md

Co-authored-by: greenie-msft <56556602+greenie-msft@users.noreply.github.com>
Signed-off-by: halspang <70976921+halspang@users.noreply.github.com>

halspang committed:
Update daprdocs/content/en/operations/resiliency/policies.md

Co-authored-by: greenie-msft <56556602+greenie-msft@users.noreply.github.com>
Signed-off-by: halspang <70976921+halspang@users.noreply.github.com>

halspang committed:
Update daprdocs/content/en/operations/resiliency/policies.md

Co-authored-by: greenie-msft <56556602+greenie-msft@users.noreply.github.com>
Signed-off-by: halspang <70976921+halspang@users.noreply.github.com>

halspang committed:
Add policy hierarchy section and expand defaults

Signed-off-by: Hal Spang <halspang@microsoft.com>

msfussell commented:
@halspang @greenie-msft Feedback
1) How does the `DefaultRetryPolicy` align with the `DaprBuiltInServiceRetries` and the other built in retries. These are not show in the call policy resolution stack?
2) Still believe that there should be at least 2 more examples for default policies especially in relation to 1 above and showing for the myriad components options which can get confusing. 
  - I suggest that the complete example https://docs.dapr.io/operations/resiliency/resiliency-overview/#complete-example-policy should be revised to include a default policy usage so this is made clear.
3) I realized that all the heading levels in this document are wrong, the should all be level 2 or 3 to build the RHS index on the page.
4) I would remove the heading here https://docs.dapr.io/operations/resiliency/policies/#policies which is not needed. 
5) Can you link to the Community call talk, so that people can watch the video?
6) Based on feedback, are there any best practices we can now provide as guidance? This document lacks anything practical or specific on when to use policies. 
7) Point 6) leads on to saying under the Components targets section here https://docs.dapr.io/operations/resiliency/targets/#components that some components can implement a retry policy (e.g. many pub/sub components) which we should at least say you should be aware of, and you can choose when to use which type of retry.  
8) Resiliency is not covered in the Concepts section, which means it tends to go undiscovered. At a minimum I would copy the Overview section up to Concept/Resiliency and then link this the other topic. See Configuration as the example here
9) Include resiliency policies into the Terminology topic https://docs.dapr.io/concepts/terminology/

halspang committed:
Add more detail around policy hierarchy

This commit also adds in a more complicated yaml example which
hopes to outline how policies interact better.

Signed-off-by: Hal Spang <halspang@microsoft.com>

halspang committed:
Add link to community call with resiliency

Signed-off-by: Hal Spang <halspang@microsoft.com>

msfussell committed:
Merge branch 'v1.9' into resiliency/generic_policies

greenie-msft committed:
Merge branch 'v1.9' into resiliency/generic_policies

greenie-msft committed:
Merge branch 'v1.9' into resiliency/generic_policies

hhunter-ms committed:
Merge branch 'v1.9' into resiliency/generic_policies

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+
+## Setting Default Policies
+
+In resiliency you can set default policies, which can have a broader scope. This is done through reserved keywords that let Dapr know when to apply the given policy. There are 3 default policies types: 
```
```suggestion
In resiliency you can set default policies, which have a broad scope. This is done through reserved keywords that let Dapr know when to apply the policy. There are 3 default policies types: 
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+- DefaultTimeoutPolicy
+- DefaultCircuitBreakerPolicy
+
+If these policies are defined, they would be used for every operation to a service, application, or component. They can also be modified to be more specific through the usage of additional keywords. The specific policies follow the following pattern, `Default%sRetryPolicy`, `Default%sTimeoutPolicy`, and `Default%sCircuitBreakerPolicy`. Where the `%s` is replaced by the new target of the policy. B ...
```
```suggestion
If these policies are defined, they are used for every operation to a service, application, or component. They can also be modified to be more specific through the appending of additional keywords. The specific policies follow the following pattern, `Default%sRetryPolicy`, `Default%sTimeoutPolicy`, and `Default%sCircuitBreakerPolicy`. Where the `%s` is replaced by a target of the policy. 

Below is a table of all possible default policy keywords and how they translate into a policy name.
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+
+If these policies are defined, they would be used for every operation to a service, application, or component. They can also be modified to be more specific through the usage of additional keywords. The specific policies follow the following pattern, `Default%sRetryPolicy`, `Default%sTimeoutPolicy`, and `Default%sCircuitBreakerPolicy`. Where the `%s` is replaced by the new target of the policy. B ...
+
+| Keyword                        | Target Operation                                     | Example Policy Name                                       |
```
```suggestion
| Keyword                        | Target Operation                                     | Policy Name                                       |
```

halspang replied:
I kept this as `Example` because it isn't a complete list of all possible names. Just an example of how that keyword applies into the names. If you still think I should remove example, let me know. Or I can add the complete list of everything, but my concern there was that it would be a large number and people may just glaze over while reading it.

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+| ConfigurationComponentOutbound | All configuration component operations.              | DefaultConfigurationComponentOutboundCircuitBreakerPolicy |
+| LockComponentOutbound          | All lock component operations.                       | DefaultLockComponentOutboundRetryPolicy                   | 
+
+### Policy Hierarchy
```
```suggestion
### Policy hierarchy resolution
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
````diff
@@ -132,4 +130,147 @@ spec:
\ No newline at end of file
+```
+
+## Setting Default Policies
````
```suggestion
## Setting default policies
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
````diff
@@ -94,7 +92,7 @@ spec:
 ```
 
-##### Override Default Retries
+## Override Default Retries
````
```suggestion
## Overriding default retries
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+
+### Policy Hierarchy
+
+Default policies are applied if the operation being executed matches the policy type and if there is no more specific policy targeting it. For each target type (app, actor, and component), the policy with the highest priority is a Named Policy, one that targets that construct specifically. If none exists, the policies are applied from most specific to most broad. 
```
```suggestion
Default policies are applied if the operation being executed matches the policy type and if there is no more specific policy targeting it. For each target type (app, actor, and component), the policy with the highest priority is a Named Policy, one that targets that construct specifically.

 If none exists, the policies are applied from most specific to most broad. 
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+
+Default policies are applied if the operation being executed matches the policy type and if there is no more specific policy targeting it. For each target type (app, actor, and component), the policy with the highest priority is a Named Policy, one that targets that construct specifically. If none exists, the policies are applied from most specific to most broad. 
+
+In the specific case of the [built-in retries]({{< ref "policies.md#Override Default Retries" >}}), default policies do not stop the built-in policies from running. In fact, both will be used but only under very specific circumstances. For service and actor invocation, the built-in retries deal specifically with issues connecting to the remote sidecar (if needed). As these are very important to t ...
```
```suggestion
#### How default policies and built-in retries work together

In the case of the [built-in retries]({{< ref "policies.md#Override Default Retries" >}}), default policies do not stop the built-in retry policies from running. Both are used together but only under specific circumstances.
 
For service and actor invocation, the built-in retries deal specifically with issues connecting to the remote sidecar (when needed). As these are important to the stability of the Dapr runtime, they are not disabled **unless** a named policy is specifically referenced for an operation. In some instances, there may be additional retries from both the built-in retry and the default retry policy, but this prevents an overly weak default policy from reducing the sidecar's availability/success rate. 
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+
+In the specific case of the [built-in retries]({{< ref "policies.md#Override Default Retries" >}}), default policies do not stop the built-in policies from running. In fact, both will be used but only under very specific circumstances. For service and actor invocation, the built-in retries deal specifically with issues connecting to the remote sidecar (if needed). As these are very important to t ...
+
+For applications, this yields:
```
```suggestion
Policy resolution hierarchy for applications, from most specific to most broad:
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+2. Default App Policies / Built-In Service Retries
+3. Default Policies / Built-In Service Retries
+
+For actors, this yields:
```
```suggestion
Policy resolution hierarchy for actors, from most specific to most broad:
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+2. Default Actor Policies / Built-In Actor Retries
+3. Default Policies / Built-In Actor Retries
+
+For components, this yields:
```
```suggestion
Policy resolution hierarchy for components, from most specific to most broad:
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+4. Default Component Policies / Built-In Actor Reminder Retries (if applicable)
+5. Default Policies / Built-In Actor Reminder Retries (if applicable)
+
+As an example, take the following system definition:
```
```suggestion
As an example, take the following solution consisting of three applications, three components and two actor types:
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
````diff
@@ -132,4 +130,147 @@ spec:
+- EventActor
+- SummaryActor
+
+```yaml
````
```suggestion
Below is policy that uses both default and named policies as applies these to the targets.

```yaml
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+        maxInterval: 60s
+        maxRetries: -1
+
+      fastRetries:
```
```suggestion
     # Named policy
      fastRetries:
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+        duration: 10ms
+        maxRetries: 3
+
+      retryForever:
```
```suggestion
     # Named policy
      retryForever:
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
````diff
@@ -132,4 +130,147 @@ spec:
+        retry: fastRetries
+```
+
+Below is an outline of which policies are used when attempting to call various members of the system.
````
```suggestion
The table below is a break down of which policies are applied when attempting to call the various targets in this solution.
```

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+
+Components:
+- Redis Pubsub: pubsub
+- Redis statestore: statestore
```
```suggestion
- Redis statestore: A key/value statestore
```

halspang replied:
This was actually the name of the component so it could be referenced in the resiliency doc. I'll try and make that more clear.

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+Components:
+- Redis Pubsub: pubsub
+- Redis statestore: statestore
+- CosmosDB Statestore: actorstore
```
```suggestion
- CosmosDB Statestore: An actor store
```

halspang replied:
This was actually the name of the component so it could be referenced in the resiliency doc. I'll try and make that more clear.

msfussell commented on daprdocs/content/en/operations/resiliency/policies.md:
```diff
@@ -132,4 +130,147 @@ spec:
+| ------------------ | ----------------------------------------------- |
+| AppA               | fastRetries                                     |
+| AppB               | retryForever                                    |
+| AppC               | DefaultAppRetryPolicy / DaprBuiltInActorRetries |
```
How do I know above that AppC uses actors? You need to say this.

halspang replied:
I'll make that clear.

msfussell commented on daprdocs/content/en/operations/resiliency/resiliency-overview.md:
```diff
@@ -164,5 +164,10 @@ spec:
 ## Related links
+
+<div class="embed-responsive embed-responsive-16by9">
+<iframe width="560" height="315" src="https://www.youtube.com/embed/7D6HOU3Ms6g?start=184" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```
Use this heading and format for the embedding. Not the use of No-Cookies

Watch [this video for how to use resiliency policies](https://youtu.be/7D6HOU3Ms6g?t=184):

<iframe width="360" height="315" src="https://www.youtube-nocookie.com/embed/sembed/7D6HOU3Ms6g?start=184" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

halspang committed:
Apply suggestions from code review

Co-authored-by: Mark Fussell <markfussell@gmail.com>
Signed-off-by: halspang <70976921+halspang@users.noreply.github.com>

greenie-msft committed:
Update resiliency-overview.md

greenie-msft committed:
Update resiliency-overview.md

hhunter-ms committed:
Merge branch 'v1.9' into resiliency/generic_policies
